TSPL.docc/LanguageGuide/Concurrency.md

```diff
@@ -564,33 +637,192 @@ the explicit parent-child relationships between tasks
+- Returning `nil` or an empty collection
+- Returning the partially completed work
+
+Downloading pictures could take a long time
+if the pictures are large or the network is slow.
+To let the user stop this work,
+without waiting for all of the tasks to complete,
+the tasks need check for cancellation and stop running if they are canceled.
+There are two ways a task can do this:
+by calling the [`Task.checkCancellation()`][] method,
+or by reading the [`Task.isCancelled`][] property.
+Calling `checkCancellation()` throws an error if the task is canceled;
+a throwing task can let that error propagate out of the task,
+stopping all of the task's work.
+This has the advantage of being simple to implement and understand.
+For more flexibility, use the `isCancelled` property,
+which lets you perform clean-up work as part of stopping the task,
+like closing network connections and deleting temporary files.
+
+[`Task.checkCancellation()`]: https://developer.apple.com/documentation/swift/task/3814826-checkcancellation
+[`Task.isCancelled`]: https://developer.apple.com/documentation/swift/task/3814832-iscancelled
+
+```
+let photos = await withTaskGroup(of: Data.self) { taskGroup in
+    let photoNames = await listPhotos(inGallery: "Summer Vacation")
+    for name in photoNames {
+        taskGroup.addTaskUnlessCancelled {
+            guard isCancelled == false else { return nil }
+            return await downloadPhoto(named: name)
+        }
+    }
+
+    var results: [Data] = []
+    for await photo in taskGroup {
+        if let photo { results.append(photo) }
+    }
+    return results
+}
+```
+
+<!-- XXX TR:
+Should the listing above add a call to taskGroup.cancelAll()
+to show that part of cancellation?
+If so, what's the best practice for where to put that?
 -->
 
-For more information about task groups,
-see [`TaskGroup`](https://developer.apple.com/documentation/swift/taskgroup).
+The code above makes several changes from the previous version:
+
+- Each task is added using the
+  `TaskGroup.addTaskUnlessCancelled(priority:operation:)` method,
+  to avoid adding new tasks after the task group has been canceled.
+
+- Each task checks for cancellation
+  before starting to download the photo.
+  If it has been canceled, the task returns `nil`.
+
+- At the end,
+  the task group skips `nil` values when collecting the results.
+  Handling cancellation by returning `nil`
+  means the task group can return a partial result ---
+  the photos that were already downloaded at the time of cancellation ---
+  instead of discarding that completed work.
+
+[`TaskGroup.addTaskUnlessCancelled(priority:operation:)`]: https://developer.apple.com/documentation/swift/taskgroup/addtaskunlesscancelled(priority:operation:)
+
+A full version of this code would include clean-up work
+as part of cancellation,
+like deleting partial downloads and closing network connections.
+
+An task that isn't part of a task group can handle cancellation
```

Comment:
That's not quite right.

Every task can do Task.isCancelled -- even Task{} and `async let` tasks.

And the `Task.withTaskCancellationHandler` is for "immediately being notified" when a task gets cancelled, instead of having to call `checkCancellation`. So the `onCancel` runs *concurrently* and *immediately* when a task gets cancelled.

You can show this with 

```
let task = Task {
  let video = await Task.withTaskCancellationHandler { } onCancel { ... }
}

// some time passes...
task.cancel()
```

where the cancellation handler runs immediately, even if we didn't have any `checkCancellation` code. This allows to "interrupt" APIs with cancellation.

Reply:
Changed in e3548f04.